src/nrtk_explorer/library/object_detector.py

```diff
@@ -77,17 +79,38 @@ def eval(
             batches[img.size][0].append(path)
             batches[img.size][1].append(img)
 
-        predictions_in_baches = [
-            zip(
-                image_ids,
-                self.pipeline(images, batch_size=batch_size),
-            )
-            for image_ids, images in batches.values()
-        ]
-
-        predictions_by_image_id = {
-            image_id: predictions
-            for batch in predictions_in_baches
-            for image_id, predictions in batch
-        }
-        return predictions_by_image_id
+        adjusted_batch_size = batch_size
+        while adjusted_batch_size >= 2:
+            try:
+                predictions_in_baches = [
+                    zip(
+                        image_ids,
+                        self.pipeline(images, batch_size=adjusted_batch_size),
+                    )
+                    for image_ids, images in batches.values()
+                ]
+
+                predictions_by_image_id = {
+                    image_id: predictions
+                    for batch in predictions_in_baches
+                    for image_id, predictions in batch
+                }
+                return predictions_by_image_id
+
+            except RuntimeError as e:
+                if "out of memory" in str(e) and adjusted_batch_size > 2:
+                    previous_batch_size = adjusted_batch_size
+                    adjusted_batch_size = max(2, adjusted_batch_size // 2)
+                    print(
+                        f"OOM (Pytorch exception {e}) due to batch_size={previous_batch_size}, setting batch_size={adjusted_batch_size}"
+                    )
+                else:
+                    raise
+
+            finally:
+                # Pytorch needs to freed its allocations outside of the exception context
+                gc.collect()
+                torch.cuda.empty_cache()
+
+        # We should never reach here
+        return None
```

Comment:
Should we throw?

Reply:
we already throw few line above in this execution path.

Reply:
maybe this is a c++ ism, maybe we can skip the return none

Reply:
Apparently mypy does not like not having a return statement even if it is legal in python. `src/nrtk_explorer/library/object_detector.py:59: error: Missing return statement  [return]`